mypy/meet.py

```diff
@@ -255,10 +255,11 @@ def visit_typeddict_type(self, t: TypedDictType) -> Type:
             for (_, l, r) in self.s.zip(t):
                 if not is_equivalent(l, r):
                     return self.default(self.s)
-            items = OrderedDict([
-                (item_name, s_item_type or t_item_type)
-                for (item_name, s_item_type, t_item_type) in self.s.zipall(t)
-            ])
+            item_list = cast(List[Tuple[str, Type]],
```

Comment:
I really don't like casts, and this one looks particularly scary given the complexity of what it's casting from. Suggestion:
```
item_list = [(item_name, s_item_type if s_item_type is not None else t_item_type)
             for (item_name, s_item_type, t_item_type) in self.s.zipall(t)]]  # type: List[str, Optional[Type]]
# Actually none of the items are None now,
# since at least one of s_item_type and t_item_type is not None
items = OrderedDict(cast(List[str, Type], items_list))
```
That's still a cast, but the comprehension gets type-checked.

Another alternative would built item_list is a good old-fashioned for-loop, with a declared type of List[str, Type], and an assert in the loop body that the item is not None.